typing/value_rec_check.ml

```diff
@@ -284,6 +301,31 @@ let classify_expression : Typedtree.expression -> sd =
             classify_expression could be extend to compute module
             shapes more precisely *)
         Not_recursive
+  and classify_module_expression env mexp : sd =
+    match mexp.mod_desc with
+    | Tmod_ident _ ->
+        Not_recursive
+    | Tmod_structure _ ->
+        Static
+    | Tmod_functor _ ->
+        Static
+    | Tmod_apply _ ->
+        Not_recursive
+    | Tmod_apply_unit _ ->
+        Not_recursive
+    | Tmod_constraint (mexp, _, _, coe) ->
+        begin match coe with
+        | Tcoerce_none -> classify_module_expression env mexp
+        | Tcoerce_structure _ ->
+            Static
+        | Tcoerce_functor _ -> Static
+        | Tcoerce_primitive _ ->
+            Misc.fatal_error "letrec: primitive coercion on a module"
+        | Tcoerce_alias _ ->
+            Misc.fatal_error "letrec: alias coercion on a module"
```

Comment:
I don't know what `Tcoerce_{primitive,alias}` are, would you mind looking it up and adding a comment in typedtree.mli for those as well?

Reply:
`Tcoerce_primitive` is when you export an external in the structure as a value in the signature. `Tcoerce_alias` is when you have a module alias in a structure that is exported as a regular module in the signature.
I'll add some docs before rebasing.

Reply:
(Both of those only make sense as sub-coercions of a `Tcoerce_structure` coercion)